tide/core/utils/stl.h

```diff
@@ -0,0 +1,83 @@
+/* conditions are met:                                               */
+/*                                                                   */
+/*   1. Redistributions of source code must retain the above         */
+/*      copyright notice, this list of conditions and the following  */
+/*      disclaimer.                                                  */
+/*                                                                   */
+/*   2. Redistributions in binary form must reproduce the above      */
+/*      copyright notice, this list of conditions and the following  */
+/*      disclaimer in the documentation and/or other materials       */
+/*      provided with the distribution.                              */
+/*                                                                   */
+/*    THIS  SOFTWARE IS PROVIDED  BY THE  UNIVERSITY OF  TEXAS AT    */
+/*    AUSTIN  ``AS IS''  AND ANY  EXPRESS OR  IMPLIED WARRANTIES,    */
+/*    INCLUDING, BUT  NOT LIMITED  TO, THE IMPLIED  WARRANTIES OF    */
+/*    MERCHANTABILITY  AND FITNESS FOR  A PARTICULAR  PURPOSE ARE    */
+/*    DISCLAIMED.  IN  NO EVENT SHALL THE UNIVERSITY  OF TEXAS AT    */
+/*    AUSTIN OR CONTRIBUTORS BE  LIABLE FOR ANY DIRECT, INDIRECT,    */
+/*    INCIDENTAL,  SPECIAL, EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES    */
+/*    (INCLUDING, BUT  NOT LIMITED TO,  PROCUREMENT OF SUBSTITUTE    */
+/*    GOODS  OR  SERVICES; LOSS  OF  USE,  DATA,  OR PROFITS;  OR    */
+/*    BUSINESS INTERRUPTION) HOWEVER CAUSED  AND ON ANY THEORY OF    */
+/*    LIABILITY, WHETHER  IN CONTRACT, STRICT  LIABILITY, OR TORT    */
+/*    (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING IN ANY WAY OUT    */
+/*    OF  THE  USE OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE    */
+/*    POSSIBILITY OF SUCH DAMAGE.                                    */
+/*                                                                   */
+/* The views and conclusions contained in the software and           */
+/* documentation are those of the authors and should not be          */
+/* interpreted as representing official policies, either expressed   */
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#ifndef STLUTILS_H
+#define STLUTILS_H
+
+#include <algorithm>
+
+/**
+ * A collection of STL utility functions.
+ */
+
+/**
+ * @return a copy of the elements from the first (sorted) container which are
+ *         not found in the second (sorted) container.
+ */
+template <class T>
+T set_difference(const T& v1, const T& v2)
+{
+    T diff;
+    std::set_difference(v1.begin(), v1.end(), v2.begin(), v2.end(),
+                        std::inserter(diff, diff.begin()));
+    return diff;
+}
+
+/**
+ * Optimal implementation of bool contains(container, value).
+ * Uses container.find() if available, else uses generic std::find().
+ */
+template <class C, typename T>
+inline auto _contains_impl(const C& c, const T& v, int)
```

Comment:
Why do you need int and long as the third argument?

Reply:
I assume it is needed for SFINAE to choose the container.find() variant when it's available, but I haven't checked in detail myself to be honest. I copied a solution provided here: https://codereview.stackexchange.com/questions/59997/contains-algorithm-for-stdvector

Reply:
Yes, otherwise there would be a conflict by the two definitions.